### ISSUE TITLE
PoC: src-expose push based

### ISSUE THREAD

**Issue by keegancsmith** [sourcegraph/sourcegraph#10045]
tl;dr a fun and technical proof of concept to try and make using src-expose 1 step and zero configuration.

### Problem

`src-expose serve` serves up an API and git endpoints which Sourcegraph uses to discover and clone repositories. However, this requires Sourcegraph to be able to reach `src-expose serve`. This adds the friction (and often confusion) of ensuring that networking is setup correctly such that it is reachable. Additionally it is difficult for us to document correctly across all our deployment environments.

### Proposal

Instead of `src-expose serve` listening on a port, it directly connects to Sourcegraph and keeps a persistent connection. Sourcegraph then uses that connection to query the API and perform git operations. This is motivated by the fact we want `src-expose` to be part of `src-cli` #9243. When part of src-cli we will get all the infra around authenticating and communicating with Sourcegraph. Additionally this will allow a magical experience where `src-expose` can also create/manage the external service configuration to connect to it.

This doesn't exist, and solving reliability and persistence for src-expose connecting to sourcegraph may prove to be too difficult/not possible. For example nginx/other load balancer may just constantly kill the connection. I would think using websockets + some abstraction over it to hide the unreliability would make sense.

**Comment by asdine** [comment]
Would something like gRPC work for this use case or would it be too heavy to add to the codebase?
IIRC it handles persistent connections, automatic reconnect, automatic retries etc. out of the box

**Comment by uwedeportivo** [comment]
You could start with http/2 which is built into the Go standard library if gRPC is too heavy for us.

**Comment by unknwon** [comment]
> Instead of src-expose serve listening on a port, it directly connects to Sourcegraph and keeps a persistent connection. Sourcegraph then uses that connection to query the API and perform git operations.

If I understand this correctly, for what we want initially, is to be able to clone the repo via "the persistent connection"?

> When part of src-cli we will get all the infra around authenticating and communicating with Sourcegraph. Additionally this will allow a magical experience where `src-expose` can also create/manage the external service configuration to connect to it.

I like this idea, very similar to what I thought about when I wanted to add arbitrary generic Git repos to Sourcegraph, and CRUD the list of repos.

**Comment by keegancsmith** [comment]
> Would something like gRPC work for this use case or would it be too heavy to add to the codebase?
> IIRC it handles persistent connections, automatic reconnect, automatic retries etc. out of the box

Yes this would be ideal, but would need to go over the internet. So we need to rely on lowest common denominator technology. EG nginx/a cloud loadbalancer with conservative settings sitting in front of sourcegraph-frontend

> You could start with http/2 which is built into the Go standard library if gRPC is too heavy for us.

See above. We can't guarantee http/2. We could always look into upgrading to using it (which would be amazing!)

> If I understand this correctly, for what we want initially, is to be able to clone the repo via "the persistent connection"?

Yes. So you need to do some fun stuff to make that work. I'm thinking repo-updater has a registry of connections it can proxy back. So when gitserver clones it actually goes `gitserver -> repo-updater -> frontend <-internet-> src-expose`

**Comment by unknwon** [comment]
> > If I understand this correctly, for what we want initially, is to be able to clone the repo via "the persistent connection"?
> 
> Yes. So you need to do some fun stuff to make that work. I'm thinking repo-updater has a registry of connections it can proxy back. So when gitserver clones it actually goes `gitserver -> repo-updater -> frontend <-internet-> src-expose`

At the core of src-expose side, it could behave like a simple Git HTTP server: 

https://github.com/gogs/gogs/blob/60273d3d6d97dd6e5220ac77a623d72fbc71e70d/internal/route/repo/http.go#L362-L372

Basically piping stdout of `git-upload-pack --stateless-rpc --advertise-refs` to the `http.ResponseWriter` and `http.ServceFile` for meta files on disk.

On the chain of Git client side (i.e. `frontend <-internet-> src-expose`), we use the "persistent connection" to send/proxy `http.Request` or equivalent as a command to ask src-expose to "upload" corresponding data with a request ID. At bare minimum, it could be as simple as src-expose making request to another HTTP endpoint (e.g. "https://sourcegraph-frontend/-/git/upload/<request ID>"), and piping `git-upload-pack --stateless-rpc --advertise-refs` and disk files as request body.

**Comment by keegancsmith** [comment]
ᴜɴᴋɴᴡᴏɴ <notifications@github.com> writes:

> At the core of src-expose side, it could behave like a simple Git HTTP server: 
>
> https://github.com/gogs/gogs/blob/60273d3d6d97dd6e5220ac77a623d72fbc71e70d/internal/route/repo/http.go#L362-L372
>
> Basically piping stdout of `git-upload-pack --stateless-rpc
> --advertise-refs` to the `http.ResponseWriter` and `http.ServceFile`
> for meta files on disk.

src-expose already can serve up git repos. See "src-expose serve". It
doesn't use "smart" http protocol like that does, instead it uses the
"dumb" protocol. However, it would be nice to start using the "smart"
protocol. For reference on smart vs dumb
https://github.com/git/git/blob/master/Documentation/technical/http-protocol.txt

FYI you could add protocolv2 support pretty easily to gogs. Take the
"Git-Protocol" HTTP header and use its value as the "GIT_PROTOCOL"
environment variable :)

> On the chain of Git client side (i.e. `frontend <-internet->
> src-expose`), we use the "persistent connection" to send/proxy
> `http.Request` or equivalent as a command to ask src-expose to
> "upload" corresponding data with a request ID. At bare minimum, it
> could be as simple as src-expose making request to another HTTP
> endpoint (e.g. "https://sourcegraph-frontend/-/git/upload/<request
> ID>"), and piping `git-upload-pack --stateless-rpc --advertise-refs`
> and disk files as request body.

I think introducing something which pushes would be a much larger change
touching many different parts. Additionally it would make our system
more complex / spread the complexity. The idea I propose above is more
about hiding that. IE git clone would hit something like
"https://repo-updater/-/src-expose-proxy/proxy-id/repo-id" and that
would then connect up with the active connection from src-expose. That
would then look like normal HTTP or even TCP, but would actually go over
websocket. Sounds a bit crazy, but we likely have ways to negotiate and
upgrade to straight forward TCP in most cases.

The way to think of this is like an SSH tunnel.

I'm not sure if I am making sense or if I am understanding your point of
view :)

**Comment by unknwon** [comment]
> src-expose already can serve up git repos. See "src-expose serve". It doesn't use "smart" http protocol like that does, instead it uses the "dumb" protocol. However, it would be nice to start using the "smart" protocol. For reference on smart vs dumb [git/git:Documentation/technical/http-protocol.txt@`master`](https://github.com/git/git/blob/master/Documentation/technical/http-protocol.txt)

Thanks for the link, will read that!

> FYI you could add protocolv2 support pretty easily to gogs. Take the "Git-Protocol" HTTP header and use its value as the "GIT_PROTOCOL" environment variable :)

Nice, didn't know that, taking notes... 📝 

> I think introducing something which pushes would be a much larger change touching many different parts. Additionally it would make our system more complex / spread the complexity. The idea I propose above is more about hiding that. IE git clone would hit something like "https://repo-updater/-/src-expose-proxy/proxy-id/repo-id" and that would then connect up with the active connection from src-expose. That would then look like normal HTTP or even TCP, but would actually go over websocket. Sounds a bit crazy, but we likely have ways to negotiate and upgrade to straight forward TCP in most cases.

Right, making "the persistent connection" behave like a `net.Conn`, sorry about my naive example 😛

**Comment by keegancsmith** [comment]
@unknwon expressed interest in doing this next iteration, so assigning him since it is directly tied to a product KR for Q2.

**Comment by unknwon** [comment]
Please see https://gist.github.com/unknwon/02ce2a8bf487dd2f2ec6946817284aeb for details of the PoC for cloning Git repo over WebSocket using "dumb" protocol.

**Comment by keegancsmith** [comment]
PoC looks great and I commented on the gist. But I would suspect we can move forward with this and get something working end to end with sourcegraph and src-expose.

**Comment by unknwon** [comment]
Close until further input.
